### PR TITLE
Add measurements to blueprints

### PR DESCRIPTION
**Pull request by labbott**

### CODE REVIEW

File: nexus/db-model/src/deployment.rs [hunk 1]
```diff
@@ -1219,6 +1221,52 @@ impl TryFrom<DbBpZoneImageSourceColumns> for BlueprintZoneImageSource {
     }
 }
 
+#[derive(Queryable, Clone, Debug, Selectable, Insertable)]
+#[diesel(table_name = bp_single_measurements)]
+//#[diesel(check_for_backend(diesel::pg::Pg))]
```
Comment (jgallagher) [inline nexus/db-model/src/deployment.rs]: Leftover debugging?

File: nexus/db-model/src/deployment.rs [hunk 2]
```diff
@@ -1219,6 +1221,52 @@ impl TryFrom<DbBpZoneImageSourceColumns> for BlueprintZoneImageSource {
...
+        BlueprintSingleMeasurement {
+            version: match artifact {
+                Some(a) => {
+                    BlueprintArtifactVersion::Available { version: a.version.0 }
+                }
+                None => BlueprintArtifactVersion::Unknown,
+            },
+            hash: *self
+                .image_artifact_sha256
+                .expect("this should always be set"),
```
Comment (jgallagher) [inline nexus/db-model/src/deployment.rs]: Can the field be non-optional (and the corresponding db row have a `NOT NULL` constraint)?

File: nexus/db-model/src/deployment.rs [hunk 3]
```diff
@@ -1219,6 +1221,52 @@ impl TryFrom<DbBpZoneImageSourceColumns> for BlueprintZoneImageSource {
...
+#[derive(Queryable, Clone, Debug, Selectable, Insertable)]
+#[diesel(table_name = bp_single_measurements)]
+//#[diesel(check_for_backend(diesel::pg::Pg))]
+pub struct BpSingleMeasurement {
+    pub blueprint_id: DbTypedUuid<BlueprintKind>,
+    pub sled_id: DbTypedUuid<SledKind>,
+    pub id: DbTypedUuid<MeasurementKind>,
+
+    pub image_artifact_sha256: Option<ArtifactHash>,
+    pub prune: bool,
```
Comment (jgallagher) [inline nexus/db-model/src/deployment.rs]: Discussed in chat but noting for completeness - I think we can drop the `prune` field since the planner has access to both the current and previous TUF repo.

File: nexus/types/src/deployment.rs [hunk 4]
```diff
@@ -1820,6 +1908,93 @@ impl fmt::Display for BlueprintArtifactVersion {
...
+    }
+}
+
+impl From<BlueprintMeasurements> for BTreeSet<BlueprintSingleMeasurement> {
+    fn from(value: BlueprintMeasurements) -> Self {
+        value.measurements
+    }
+}
+
+impl From<BTreeSet<BlueprintSingleMeasurement>> for BlueprintMeasurements {
```
Comment (jgallagher) [inline nexus/types/src/deployment.rs]: These `From`s seem fine-ish but surprising - could wherever we're currently using `BTreeSet<BlueprintSingleMeasurement>` use `BlueprintMeasurements` instead?

File: nexus/types/src/deployment.rs [hunk 5]
```diff
@@ -1820,6 +1908,93 @@ impl fmt::Display for BlueprintArtifactVersion {
...
+impl BlueprintMeasurements {
+    pub fn default_contents() -> Self {
+        Self { measurements: BTreeSet::new() }
+    }
+
+    pub fn append_measurement(&mut self, single: BlueprintSingleMeasurement) {
+        self.measurements.insert(single);
+    }
+
+    // An empty measurement set here corresponds to the install dataset
```
Comment (jgallagher) [inline nexus/types/src/deployment.rs]: Is this property guaranteed? Wondering if there are edge cases where we could have an empty set of measurements but that didn't imply "use the install dataset" - maybe dev/test systems? Or if this is a surprising security-related property in the future sometime if someone assumes "empty set" means "won't be able to talk" but it actually means "falls back to whatever was most recently written to the install dataset (possibly months/years ago)".

Comment (labbott) [reply nexus/types/src/deployment.rs]: Yeah this is a bit of a strange edge case. I did a first pass with this being an enum like `BlueprintZoneImageSource` but I'm trying to remember why doing that here was ugly. It's possible I went a layer too far in getting rid of indirection. I'll play around with this more.

That said, there eventually should be no case in the future when this is empty past the initial install. As we get further along and more confident I expect we'll want to make that a stronger assertion.

Comment (labbott) [reply nexus/types/src/deployment.rs]: This turned out to be more doable than I remembered, perhaps now that things are matured

File: nexus/types/src/deployment.rs [hunk 6]
```diff
@@ -967,6 +969,68 @@ impl BpTableData for BlueprintHostPhase2TableData<'_> {
...
+                vec![BpTableRow::from_strings(
+                    state,
+                    vec![
+                        "install dataset".to_string(),
+                        "(no version)".to_string(),
+                    ],
+                )]
+                .into_iter()
+                .collect::<Vec<BpTableRow>>()
+                .into_iter()
```
Comment (labbott) [inline nexus/types/src/deployment.rs]: I need to fix this up to use `Either`

File: nexus/types/src/deployment.rs [hunk 7]
```diff
@@ -1820,6 +1897,94 @@ impl fmt::Display for BlueprintArtifactVersion {
...
+            BlueprintMeasurements::InstallDataset => {
+                std::iter::empty::<&BlueprintSingleMeasurement>()
+                    .collect::<Vec<&BlueprintSingleMeasurement>>()
+                    .into_iter()
+            }
+            BlueprintMeasurements::Artifacts { artifacts } => artifacts
+                .iter()
+                .collect::<Vec<&BlueprintSingleMeasurement>>()
+                .into_iter(),
+        }
```
Comment (labbott) [inline nexus/types/src/deployment.rs:2195]: Fix this to use `Either`

File: nexus/db-queries/src/db/datastore/deployment.rs [hunk 8]
```diff
@@ -1229,6 +1264,76 @@ impl DataStore {
             );
         }
 
+        let raw_measurements: Vec<(BpSingleMeasurement, Option<TufArtifact>)> = {
```
Comment (jgallagher) [inline nexus/db-queries/src/db/datastore/deployment.rs]: I think this is in the wrong place within this function, after the changes @smklein made to ensure we don't read partially-deleted blueprints (#9703). About 50 lines above this is:

```
        // ================================================================
        // STAGE 3: Parse and validate all loaded data
        //
        // At this point, we know the blueprint exists (wasn't deleted).
        // Any validation errors from here indicate real database
        // corruption, not a torn read from concurrent deletion.
        // ================================================================
```

and we're not expected to do any more reads from the DB in this stage, so I think this needs to go above into "STAGE 1". (Maybe we should break the stages up into three methods to make this mistake more difficult? But that probably has its own annoyances.)

I think this may also mean we don't need the

```
                // We start with measurements from the install dataset, if
                // there's anything in the database this will get overwritten
```

comment above - if we've already read `raw_measurements` before we get to that point, hopefully we can fill in the correct value immediately?

File: nexus/types/src/deployment.rs [hunk 9]
```diff
@@ -1364,6 +1439,8 @@ pub struct BlueprintSledConfig {
     pub zones: IdOrdMap<BlueprintZoneConfig>,
     pub remove_mupdate_override: Option<MupdateOverrideUuid>,
     pub host_phase_2: BlueprintHostPhase2DesiredSlots,
+    #[serde(default = "BlueprintMeasurements::default")]
```
Comment (jgallagher) [inline nexus/types/src/deployment.rs]: Do we need a `#[serde(default)]` here? I don't think we need to parse old `BlueprintSledConfig`s.

File: nexus/types/src/deployment/blueprint_display.rs [hunk 10]
```diff
@@ -352,6 +352,18 @@ impl fmt::Display for BpTable {
     }
 }
 
+/// The [`BpTable`] schema for desired host phase 2 contents
```
Comment (jgallagher) [inline nexus/types/src/deployment/blueprint_display.rs]: Copy/pasta typo

File: schema/crdb/dbinit.sql [hunk 11]
```diff
@@ -5195,6 +5195,18 @@ CREATE TABLE IF NOT EXISTS omicron.public.bp_sled_metadata (
...
+-- description of measurements specified in a blueprint
+CREATE TABLE IF NOT EXISTS omicron.public.bp_single_measurements (
+    -- foreign key into the `blueprint` table
+    blueprint_id UUID NOT NULL,
+    sled_id UUID NOT NULL,
+    -- id solely for database purposes
+    id UUID NOT NULL,
+
+     image_artifact_sha256 STRING(64) NOT NULL,
+     PRIMARY KEY (blueprint_id, id)
```
Comment (jgallagher) [inline schema/crdb/dbinit.sql]: Turbo-nit - these two lines are indented differently from the columns above

File: nexus/types/src/deployment.rs (hunk 9 above)
Comment (labbott) [reply nexus/types/src/deployment.rs]: This was something I added when I thought I had a different problem. I can remove it.

File: nexus/db-queries/src/db/datastore/deployment.rs (hunk 8 above)
Comment (labbott) [reply nexus/db-queries/src/db/datastore/deployment.rs]: Good catch. I'll see about making this change and mayyybe see about refactoring but I'm not great with diesel. Maybe I should just file a bug instead?

File: nexus/db-model/src/schema_versions.rs [hunk 12]
```diff
@@ -16,7 +16,7 @@ use std::{collections::BTreeMap, sync::LazyLock};
 ///
 /// This must be updated when you change the database schema.  Refer to
 /// schema/crdb/README.adoc in the root of this repository for details.
-pub const SCHEMA_VERSION: Version = Version::new(231, 0, 0);
+pub const SCHEMA_VERSION: Version = Version::new(232, 0, 0);
```
Comment (jgallagher) [inline nexus/db-model/src/schema_versions.rs]: We have to hold off on merging this until 18.1 is cut (should be soon...)

File: nexus/db-queries/src/db/datastore/deployment.rs [hunk 13]
```diff
@@ -1229,6 +1307,31 @@ impl DataStore {
...
+            omicron_measurements
+                .entry(m.sled_id.into())
+                .or_default()
+                .insert(m.to_measurement(artifact));
+        }
+
+        for (id, artifacts) in omicron_measurements {
+            let sled_config = sled_configs.get_mut(&id).ok_or_else(|| {
+                // This means we found a measurement without
+                // the associated sled
```
Comment (jgallagher) [inline nexus/db-queries/src/db/datastore/deployment.rs]: This confirms we don't have unexpected measurement rows that don't match any sled configs (good) - do we need to check the other way around (i.e., that any sled config has measurements)? Or is "no measurement rows" the implicit db representation for `BlueprintMeasurements::InstallDataset`?

File: nexus/types/src/deployment.rs [hunk 14]
```diff
@@ -967,6 +969,85 @@ impl BpTableData for BlueprintHostPhase2TableData<'_> {
...
+            }
+            BlueprintMeasurements::InstallDataset => either::Either::Right(
+                vec![BpTableRow::from_strings(
+                    state,
+                    vec![
+                        "install dataset".to_string(),
+                        "(no version)".to_string(),
+                    ],
+                )]
+                .into_iter(),
```
Comment (jgallagher) [inline nexus/types/src/deployment.rs]: Tiny nit - I think this outer `vec![...].into_iter()` could be `std::iter::once(...)` instead, both to save a small heap allocation and for clarity?

File: nexus/types/src/deployment.rs [hunk 15]
```diff
@@ -1820,6 +1913,120 @@ impl fmt::Display for BlueprintArtifactVersion {
...
+            ) => first.diff(second),
+        }
+    }
+}
+
+impl BlueprintMeasurements {
+    fn len(&self) -> usize {
+        match self {
+            BlueprintMeasurements::InstallDataset => 0,
+            BlueprintMeasurements::Artifacts { artifacts } => artifacts.len(),
```
Comment (jgallagher) [inline nexus/types/src/deployment.rs]: This is maybe overly paranoid, but: is it legal to have `BlueprintMeasurements::Artifacts { artifacts }` where the inner `artifacts` set is itself empty? (If it's not, I'm wondering if it's worth the ergonomics pain to change this to some kind of `NonEmptyBTreeSet<BlueprintSingleMeasurement>` instead.)

File: nexus/types/src/deployment.rs [hunk 16]
```diff
@@ -1820,6 +1913,120 @@ impl fmt::Display for BlueprintArtifactVersion {
...
+        }
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = &BlueprintSingleMeasurement> {
+        match self {
```
Comment (jgallagher) [inline nexus/types/src/deployment.rs]: I think this would be clearer with `Either`:

```rust
        match self {
            BlueprintMeasurements::InstallDataset => {
                either::Either::Left(std::iter::empty())
            }
            BlueprintMeasurements::Artifacts { artifacts } => {
                either::Either::Right(artifacts.iter())
            }
        }
```

File: schema/crdb/dbinit.sql [hunk 17]
```diff
@@ -5213,6 +5213,18 @@ CREATE TABLE IF NOT EXISTS omicron.public.bp_sled_metadata (
     PRIMARY KEY (blueprint_id, sled_id)
 );
 
+-- description of measurements specified in a blueprint
+CREATE TABLE IF NOT EXISTS omicron.public.bp_single_measurements (
+    -- foreign key into the `blueprint` table
+    blueprint_id UUID NOT NULL,
+    sled_id UUID NOT NULL,
+    -- id solely for database purposes
```
Comment (jgallagher) [inline schema/crdb/dbinit.sql]: Two questions:

* What are the database purposes?
* (Maybe depends on the details of the former) What prevents `id` from being the sole primary key?

File: nexus/db-model/src/deployment.rs [hunk 18]
```diff
@@ -1219,6 +1221,46 @@ impl TryFrom<DbBpZoneImageSourceColumns> for BlueprintZoneImageSource {
...
+impl BpSingleMeasurement {
+    pub fn new(
+        blueprint_id: BlueprintUuid,
+        sled_id: SledUuid,
+        measurement: &BlueprintSingleMeasurement,
+    ) -> Self {
+        Self {
+            blueprint_id: blueprint_id.into(),
+            sled_id: sled_id.into(),
+            id: omicron_uuid_kinds::MeasurementUuid::new_v4().into(),
```
Comment (andrewjstone) [inline nexus/db-model/src/deployment.rs]: Nit, I'd probably import `MeasurementUuid` like the other typed UUIDs.

File: nexus/db-queries/src/db/datastore/deployment.rs (hunk 13 above)
Comment (labbott) [reply nexus/db-queries/src/db/datastore/deployment.rs]: no measurement rows implies `BlueprintMeasurements::InstallDataset`. This gets set as the default entry and gets overwritten when we pull things out.

File: nexus/types/src/deployment.rs [hunk 19]
```diff
@@ -967,6 +969,85 @@ impl BpTableData for BlueprintHostPhase2TableData<'_> {
...
+/// Wrapper to display a table of a `BlueprintSledConfig`'s measurements.
+#[derive(Clone, Debug)]
+struct BlueprintMeasurementsTableData<'a> {
+    measurements: &'a BlueprintMeasurements,
+}
+
+impl<'a> BlueprintMeasurementsTableData<'a> {
+    fn diff_rows<'b>(
+        diffs: &'b daft::BTreeSetDiff<BlueprintSingleMeasurement>,
+    ) -> impl Iterator<Item = BpTableRow> + 'b + use<'b> {
```
Comment (andrewjstone) [inline nexus/types/src/deployment.rs:981]: TIL about `use` syntax. I don't think you need the `'b` if you already have the `use<'b>`.

File: nexus/db-queries/src/db/datastore/deployment.rs (hunk 13 above)
Comment (jgallagher) [reply nexus/db-queries/src/db/datastore/deployment.rs]: We have some checks in the external API when an operator tries to start an update for "has a mupdate taken place". Do we need to update those checks to look for this value too? That doesn't seem quite right if we're going to default existing blueprints to `InstallDataset`.

File: nexus/types/src/deployment.rs (hunk 19 above)
Comment (andrewjstone) [inline nexus/types/src/deployment.rs:981]: I'm slightly weirded out by the use of `Either` here to return an interator rather than just returning the underlying vecs.

File: nexus/types/src/deployment.rs [hunk 20]
```diff
@@ -1820,6 +1913,120 @@ impl fmt::Display for BlueprintArtifactVersion {
...
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize, JsonSchema, PartialEq, Eq)]
+#[serde(tag = "type", rename_all = "snake_case")]
+pub enum BlueprintMeasurements {
+    InstallDataset,
+    Artifacts { artifacts: BTreeSet<BlueprintSingleMeasurement> },
+}
+
+// Enums will get treated as a `Leaf` which doesn't product correct
```
Comment (andrewjstone) [inline nexus/types/src/deployment.rs]:
```suggestion
// Enums will get treated as a `Leaf` which doesn't produce correct
```

File: nexus/types/src/deployment.rs (hunk 19 above)
Comment (jgallagher) [reply nexus/types/src/deployment.rs:981]: Oh, this is the same as https://github.com/oxidecomputer/omicron/pull/9718#discussion_r2823019285. For the `Left` case, there isn't an underlying vec - it should be using `std::iter::once()` instead.

File: schema/crdb/dbinit.sql (hunk 17 above)
Comment (labbott) [reply schema/crdb/dbinit.sql]: This was me both doing some copy/paste and running into diesel pain. Most other parts of the system like disks/datasets already have an id. I ran into some diesel headaches when I tried it without the `id`.  I think we always want to be looking up for a specific blueprint so having that be part of the primary key made sense to me.

File: nexus/types/src/deployment.rs (hunk 15 above)
Comment (labbott) [reply nexus/types/src/deployment.rs]: It shouldn't be legal. This is one of the awkward spots about the enum modeling vs just a struct that I don't love. I see https://docs.rs/nonempty/latest/nonempty/ for `vec` but I don't see anything like that for sets.  I guess this could be an opaque inner type and we could check at creation time and panic if we try and create something with zero length?  It's also worth considering the cases where we could end up constructing this right now are pretty narrow.

Comment (jgallagher) [reply nexus/types/src/deployment.rs]: > I guess this could be an opaque inner type and we could check at creation time and panic if we try and create something with zero length?

Yeah, this is the direction I'd go (although I wouldn't panic: I'd either return an error or make it return an `Option` like the `NonZero*` integer types in the std lib, since there's exactly one input that can possibly fail).

> It's also worth considering the cases where we could end up constructing this right now are pretty narrow.

If there aren't many ways to construct this, it seems like it might be worth the work to enforce that construction guarantees at least one artifact?

File: dev-tools/reconfigurator-cli/tests/output/cmds-add-sled-no-disks-stdout [hunk 21]
```diff
@@ -71,6 +71,13 @@ parent:    dbcbd3d6-41ff-48ae-ac0b-1becc9b2fd21
     B      current contents 
 
 
+    measurements:
+    ---------------------------
+    hash      version          
+    ---------------------------
+    unknown   (unknown version)
```
Comment (jgallagher) [inline dev-tools/reconfigurator-cli/tests/output/cmds-add-sled-no-disks-stdout:78]: I only skimmed all these expectorate outputs, but just checking: it's expected that some will have `unknown version` and some will have `no version`?

File: nexus/types/src/deployment.rs [hunk 22]
```diff
@@ -967,6 +969,200 @@ impl BpTableData for BlueprintHostPhase2TableData<'_> {
...
+        match diffs {
+            BlueprintMeasurementsDiff::NoInstallChange => {
+                std::iter::once(BpTableRow::from_strings(
+                    BpDiffState::Unchanged,
+                    vec![
+                        "install dataset".to_string(),
+                        "(no version)".to_string(),
+                    ],
+                ))
+                .collect::<Vec<_>>()
```
Comment (jgallagher) [inline nexus/types/src/deployment.rs]: I don't think collecting an `iter::once()` into a single-element `Vec` is necessary; we can construct the vec directly:

```rust
vec![BpTableRow::from_strings(...args...)].into_iter()
```

(same note in the next branch)

File: nexus/types/src/deployment.rs [hunk 23]
```diff
@@ -967,6 +969,200 @@ impl BpTableData for BlueprintHostPhase2TableData<'_> {
...
+                BpTableRow::from_strings(
+                    BpDiffState::Removed,
+                    vec!["unknown".to_string(), "(no version)".to_string()],
+                ),
+            )
+            .chain(std::iter::once(BpTableRow::from_strings(
+                BpDiffState::Added,
+                vec!["install dataset".to_string(), "(no version)".to_string()],
+            )))
+            .collect::<Vec<_>>()
```
Comment (jgallagher) [inline nexus/types/src/deployment.rs]: Same note on this one - this would be clearer as

```rust
vec![
    BpTableRow::from_strings(...the removed row...),
    BpTableRow::from_strings(...the added row...),
].into_iter()
```

File: nexus/types/src/deployment.rs [hunk 24]
```diff
@@ -1820,6 +2028,259 @@ impl fmt::Display for BlueprintArtifactVersion {
...
+
+impl Display for BlueprintSingleMeasurement {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{} {}", self.hash, self.version)
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize, JsonSchema, PartialEq, Eq)]
+#[serde(tag = "type", rename_all = "snake_case")]
+pub enum BlueprintMeasurementsInternal {
```
Comment (jgallagher) [inline nexus/types/src/deployment.rs]: We shouldn't derive `Deserialize` - that would allow construction of a `BlueprintMeasurementsInternal::Artifacts` with an empty set of artifacts via a JSON blob with the right tag/contents. We need to manually implement it to reject those. (We have various examples of this floating around - usually we have a deserialize impl that internally has a type with the same shape as this type, but then validates the cases that need extra validation.)

That said, I don't think this is how I'd split the public/private details. It means callers are forced to do an `if is_install { ... } else if is_unknown { ... } else { ... }` chain instead of matching; what if we split it like this:

```rust
// clients can match and not use if/else chains
pub enum BlueprintMeasurements {
    Unknown,
    InstallDataset,
    Artifacts { artifacts: BlueprintArtifactMeasurements },
}

// private inner field; construction guarantees at least one measurement in the inner set
pub struct BlueprintArtifactMeasurements(BTreeSet<BlueprintSingleMeasurement>);
```

which pushes the type safety down as far as it can go. (The `BlueprintArtifactMeasurements` struct here would still need a non-derived `Deserialized` for the same reasons noted above.)

File: schema/crdb/dbinit.sql [hunk 25]
```diff
@@ -5210,9 +5217,24 @@ CREATE TABLE IF NOT EXISTS omicron.public.bp_sled_metadata (
         CHECK (last_allocated_ip_subnet_offset BETWEEN 0 AND 65535)
         NOT NULL,
 
+    -- the measurements for this sled
+    measurements omicron.public.bp_sled_measurements NOT NULL DEFAULT 'unknown',
```
Comment (jgallagher) [inline schema/crdb/dbinit.sql]: We generally stay away from `DEFAULT`s on columns. We need it in the migration to populate existing rows, but then a followup migration step can alter the column and drop `DEFAULT`.

File: schema/crdb/dbinit.sql (hunk 17 above)
Comment (jgallagher) [reply schema/crdb/dbinit.sql]: Should we revisit this? Either address whatever diesel pain caused this column to exist and figure out how to drop it, or expand the comment to explain why it's here?

File: dev-tools/reconfigurator-cli/tests/output/cmds-add-sled-no-disks-stdout (hunk 21 above)
Comment (labbott) [reply dev-tools/reconfigurator-cli/tests/output/cmds-add-sled-no-disks-stdout:78]: I think I forgot to update a few tests but `install` gets  `no version` vs `unknown version`, I'll double check to make sure these are consistent

File: nexus/types/src/deployment.rs (hunk 24 above)
Comment (labbott) [reply nexus/types/src/deployment.rs]: ah! not being able to match was very annoying. let me play around with this.

File: schema/crdb/dbinit.sql (hunk 17 above)
Comment (labbott) [reply schema/crdb/dbinit.sql]: Okay I think this needs to stay with a bigger explanation. We store measurements per sled. So a measurement with a single hash needs to exist in the database for all sleds in the same blueprint. This means we need some sort of ID for it to be a unique entry.

Comment (labbott) [reply schema/crdb/dbinit.sql]: The correct answer is just have the primary key be `(blueprint_id, sled_id, hash)`

File: nexus/db-queries/src/db/datastore/deployment.rs [hunk 26]
```diff
@@ -1208,6 +1299,42 @@ impl DataStore {
...
+                    }
+                }
+                // There were no measurements, we expect this for both these cases
+                (None, DbBpSledMeasurements::InstallDataset) => {
+                    BlueprintMeasurements::InstallDataset
+                }
+                (None, DbBpSledMeasurements::Unknown) => {
+                    BlueprintMeasurements::Unknown
+                }
+                // The rest of these are inconsistent and we have a bug somewhere
```
Comment (jgallagher) [inline nexus/db-queries/src/db/datastore/deployment.rs:1323]: Thanks for all the rework here; the explicit `DbBpSledMeasurements` feels much more solid than having to infer it from the presence or absence of other rows.

File: nexus/types/src/deployment.rs [hunk 27]
```diff
@@ -967,6 +969,212 @@ impl BpTableData for BlueprintHostPhase2TableData<'_> {
     }
 }
 
+/// Wrapper to display a table of a `BlueprintSledConfig`'s measurements.
+#[derive(Clone, Debug)]
+struct BlueprintMeasurementsTableData<'a> {
+    measurements: &'a BlueprintMeasurements,
+}
+
+impl<'a> BlueprintMeasurementsTableData<'a> {
```
Comment (andrewjstone) [inline nexus/types/src/deployment.rs:978]: I would love a mechanism to auto-generate these tables so everyone didn't have to do this by hand when handed a diff. I'm not sure it's totally feasible, but maybe one day I'll get back to it.

File: nexus/types/src/deployment.rs [hunk 28]
```diff
@@ -967,6 +969,212 @@ impl BpTableData for BlueprintHostPhase2TableData<'_> {
...
+    fn diff_rows<'b>(
+        diffs: &'b BlueprintMeasurementsDiff,
+    ) -> impl Iterator<Item = BpTableRow> + 'b + use<'b> {
+        match diffs {
+            BlueprintMeasurementsDiff::NoInstallChange => {
+                vec![BpTableRow::from_strings(
+                    BpDiffState::Unchanged,
+                    vec![
+                        "install dataset".to_string(),
```
Comment (andrewjstone) [inline nexus/types/src/deployment.rs:987]: nit: I probably would have declared constants for these common strings, but it's not a big deal. They appear to match and get used correctly from my review.

File: nexus/types/src/deployment.rs (hunk 27 above)
Comment (labbott) [reply nexus/types/src/deployment.rs:978]: :smiling_face_with_tear:
